include/native/pipeline.h

```diff
@@ -21,398 +21,514 @@
+          internal::requires_index_not_last<Index,MoreTransformers...> = 0>
+void composed_pipeline(InQueue & input_queue, 
+                       const pipeline_info<parallel_execution_native,MoreTransformers...> & pipeline_obj, 
+                       OutQueue & output_queue, std::vector<std::thread> & tasks)
 {
-      typedef typename std::tuple_element<currentStage, decltype(pipe.stages)>::type lambdaType;
-//      typedef typename std::remove_reference<decltype(lambdaPointerType())>::type  lambdaType; 
-      typedef typename std::result_of< lambdaType (typename InStream::value_type::value_type) > ::type queueType;
-
-
-
-      static mpmc_queue<std::experimental::optional<queueType>> queueOut(pipe.exectype.queue_size,pipe.exectype.lockfree); 
-
-      composed_pipeline(pipe.exectype, qin, std::get<currentStage>(pipe.stages), queueOut, tasks);
-      composed_pipeline<mpmc_queue<std::experimental::optional<queueType>>,OutStream, currentStage+1, Stages ...>(queueOut,pipe,qout,tasks);
-       
+  using stage_type = 
+      typename std::tuple_element<Index,decltype(pipeline_obj.stages)>::type;
+  using input_type = typename InQueue::value_type;
+  using input_value_type = typename input_type::value_type;
+  using result_type = 
+      typename std::result_of<stage_type(input_type)>::type;
+
+  // TODO: Why static?
+  static mpmc_queue<std::experimental::optional<result_type>> tmp_queue{
+      pipeline_obj.exectype.queue_size, pipeline_obj.exectype.lockfree}; 
+
+  composed_pipeline(pipeline_obj.exectype, input_queue, 
+      std::get<Index>(pipeline_obj.stages), tmp_queue, tasks);
+  composed_pipeline<mpmc_queue<std::experimental::optional<result_type>>, 
+      OutQueue, Index+1, MoreTransformers ...>(
+          tmp_queue,pipeline_obj,output_queue,tasks);
 }
 
-template <typename InStream, typename Stage, typename OutStream>
- void composed_pipeline(parallel_execution_native &p, InStream &qin, Stage const & s, OutStream &qout, std::vector<std::thread> & tasks){
-    tasks.push_back(
-      std::thread([&](){
-        using lambdaType = Stage;
-        p.register_thread();
+template <typename InQueue, typename Transformer, typename OutQueue>
+void composed_pipeline(parallel_execution_native & ex, InQueue & input_queue,
+                       Transformer && transform_op, OutQueue & output_queue, 
+                       std::vector<std::thread> & tasks)
+{
+  using namespace std;
+  tasks.push_back(
+    thread{[&](){
+      ex.register_thread();
         
-        auto item = qin.pop();
-        while(true){
-           if(!item) {
-                qout.push( typename OutStream::value_type::value_type () );
-                break;
-           }else{
-                auto out = typename OutStream::value_type::value_type( (*s)(item.value()) );
-                qout.push(out);
-           }
-           item = qin.pop();
+      auto item = input_queue.pop();
+      for (;;) {
+        using output_type = typename OutQueue::value_type;
```

Comment:
I found a bug in the composed pipeline. Instead of pushing elements of value_type::value_type of the queue (i.e. the actual element type) it should introduce the value_type of the queue (i.e. the optional type).